### PR TITLE
MG-2080 - Add TLS Certs Loading and Certificate Revocation 

### PR DESCRIPTION
**Pull request by arvindh123**
Pull request title should be `MF-XXX - description` or `NOISSUE - description` where XXX is ID of issue that this PR relate to.
Please review the [CONTRIBUTING.md](https://github.com/absmach/magistrala/blob/master/CONTRIBUTING.md) file for detailed contributing guidelines.

### What does this do?
Add TLS Certs Loading and Certificate Revocation 

### Which issue(s) does this PR fix/relate to?
Resolves https://github.com/absmach/magistrala/issues/2080.

### Notes

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 build
 test/aedes/node_modules
+ssl/certs/*
```
Comment (dborovcanin) [inline .gitignore:3]: Add the empty line.

File: examples/client/http/with_mtls.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+curl -sSiX POST "${protocol}://${host}:${port}/${path}" -H "content-type:${content}" -H "Authorization:TOKEN" -d "${message}" --cacert $cafile  --cert $reovokedcertfile --key $reovokedkeyfile
+
+echo -e "\nPosting message to ${protocol}://${host}:${port}/${path} with tls, Authorization header, ca certificates ${cafile} & unknown client certificate ${unknowncertfile} ${unknownkeyfile}..."
+curl -sSiX POST "${protocol}://${host}:${port}/${path}" -H "content-type:${content}" -H "Authorization:TOKEN" -d "${message}" --cacert $cafile  --cert $unknowncertfile --key $unknownkeyfile
+
+echo -e "\nPosting message to ${protocol}://${host}:${port}/${path} with tls, Authorization header, ca certificate ${cafile} & without client certificates.."
+curl -sSiX POST "${protocol}://${host}:${port}/${path}" -H "content-type:${content}" -H "Authorization:TOKEN" -d "${message}" --cacert $cafile 2>&1
+
+echo -e "\nPosting message to ${protocol}://${host}:${port}/${path} with tls, Authorization header,  & without ca , client certificates.."
+curl -sSiX POST "${protocol}://${host}:${port}/${path}" -H "content-type:${content}" -H "Authorization:TOKEN" -d "${message}" 2>&1
```
Comment (dborovcanin) [inline examples/client/http/with_mtls.sh:39]: Same here.

File: examples/client/http/without_tls.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+echo "Posting message to ${protocol}://${host}:${port}/${path} without tls ..."
+curl -sSiX POST "${protocol}://${host}:${port}/${path}" -H "content-type:${content}" -H "Authorization:TOKEN" -d "${message}"
+
+
+echo -e "\nPosting message to ${protocol}://${host}:${port}/${path} without tls and with basic authentication..."
+curl -sSi -u username:password -X  POST "${protocol}://${host}:${port}/${path}" -H "content-type:${content}" -d "${message}"
+
+
+echo -e "\nPosting message to invalid path ${protocol}://${host}:${port}/${invalidPath} without tls..."
+curl -sSiX POST "${protocol}://${host}:${port}/${invalidPath}" -H "content-type:${content}" -H "Authorization:TOKEN" -d "${message}"
```
Comment (dborovcanin) [inline examples/client/http/without_tls.sh:19]: And here.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (arvindh123) [reply .gitignore]: done

File: examples/client/http/without_tls.sh (hunk 3 above)
Comment (arvindh123) [reply examples/client/http/without_tls.sh]: done

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Please update Readme with details on how to use examples and create relevant certificates using Makefile.

File: config.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+type Config struct {
+	Address    string `env:"ADDRESS"                        envDefault:""`
+	Target     string `env:"TARGET"                         envDefault:""`
+	PrefixPath string `env:"PREFIX_PATH"                    envDefault:""`
+	TLSConfig  mptls.Config
+}
+
+func (c *Config) EnvParse(opts env.Options) error {
+	if len(opts.FuncMap) == 0 {
```
Comment (dborovcanin) [inline config.go:22]: Use switch instead of `if-else`.

File: pkg/tls/tls.go [hunk 5]
```diff
@@ -4,39 +4,175 @@
...
+	ClientCAFile                        string           `env:"CLIENT_CA_FILE"                             envDefault:""`
+	ClientCertValidationMethods         []ValidateMethod `env:"CLIENT_CERT_VALIDATION_METHODS"             envDefault:""`
+	OCSPDepth                           uint             `env:"OCSP_DEPTH"                                 envDefault:"0"`
+	OCSPResponderURL                    url.URL          `env:"OCSP_RESPONDER_URL"                         envDefault:""`
+	CRLDepth                            uint             `env:"CRL_DEPTH"                                  envDefault:"1"`
+	OfflineCRLFile                      string           `env:"OFFLINE_CRL_FILE"                           envDefault:""`
+	CRLDistributionPoints               url.URL          `env:"CRL_DISTRIBUTION_POINTS"                    envDefault:""`
+	CRLDistributionPointsSignCheck      bool             `env:"CRL_DISTRIBUTION_POINTS_SIGN_CHECK"         envDefault:"false"`
+	CRLDistributionPointsIssuerCertFile string           `env:"CRL_DISTRIBUTION_POINTS_ISSUER_CERT_FILE "  envDefault:""`
```
Comment (dborovcanin) [inline pkg/tls/tls.go:120]: Extract these (revocation config) to a separate struct in a separate file.

File: pkg/tls/tls.go [hunk 6]
```diff
@@ -4,39 +4,175 @@
...
+}
+
+func (c *Config) EnvParse(opts env.Options) error {
+	return env.ParseWithOptions(c, opts)
+}
+
+// Load return a TLS configuration that can be used in TLS servers.
+func (c *Config) Load() (*tls.Config, Security, error) {
```
Comment (dborovcanin) [inline pkg/tls/tls.go:128]: Let's not return `Security` here. Returning `nil, nil` to indicate no config and no error seems a little cryptic (while correct, so I would suggest using the `ErrNoTLS` error to indicate no error, but also no TLS; so you can later have something like:
```
	l, err := net.Listen("tcp", p.config.Address)
	if err != nil {
		return err
	}
	tlsCfg, err := p.config.TLSConfig.Load()
	switch err {
	case nil:
		l = tls.NewListener(l, tlsCfg)
	case mptls.ErrNoTLS:
		break
	default:
		return err
	}
```
While this means we first start the listener and then parse the TLS config, this order is not so important; we can't have one without the other anyways.

File: pkg/tls/tls.go [hunk 7]
```diff
@@ -59,3 +195,338 @@ func ClientCert(conn net.Conn) (x509.Certificate, error) {
...
 }
+
+func loadCertFile(certFile string) ([]byte, error) {
+	if certFile != "" {
+		return os.ReadFile(certFile)
+	}
+	return []byte{}, nil
+}
+
+func (c *Config) verifyPeerCertificate(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
```
Comment (dborovcanin) [inline pkg/tls/tls.go:206]: Move this alongside the new file for revocation config.

File: pkg/tls/tls.go [hunk 8]
```diff
@@ -59,3 +195,338 @@ func ClientCert(conn net.Conn) (x509.Certificate, error) {
...
+}
+
+func (c *Config) ocspVerify(peerCertificate, issuerCert *x509.Certificate) error {
+	opts := &ocsp.RequestOptions{Hash: crypto.SHA256}
+	var err error
+
+	if !isRootCA(peerCertificate) {
+		if issuerCert == nil {
+			if len(peerCertificate.IssuingCertificateURL) < 1 {
+				return fmt.Errorf("neither the issuer certificate is present in the chain nor is the issuer certificate URL present in AIA, certificate common name %s  and serial number %x", peerCertificate.Subject.CommonName, peerCertificate.SerialNumber)
```
Comment (dborovcanin) [inline pkg/tls/tls.go:311]: Extract large strings to constants. Try to simplify the method.

File: pkg/tls/tls.go (hunk 7 above)
Comment (dborovcanin) [reply pkg/tls/tls.go:206]: Also, this method looks very complex. Can you try to simplify or, at least, split?

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: pkg/tls/tls.go [hunk 9]
```diff
@@ -9,34 +9,117 @@ import (
...
+	case WithmTLS:
+		return "with mTLS"
+	case WithmTLSVerify:
+		return "with mTLS and validation of client certificate revocation status"
+	case WithoutTLS:
+		fallthrough
+	default:
+		return "without TLS"
+	}
+}
```
Comment (dborovcanin) [inline pkg/tls/tls.go:46]: Let's remove this whole construct and replace it with:
```go
func Status(c Config) string {
	if c.CertFile == "" && c.KeyFile == "" {
		return "TLS"
	}

	if c.ClientCAFile != "" {
		if len(c.ClientValidation.ValidationMethods) > 0 {
			return "mTLS with client verification"
		}
		return "mTLS"
	}

	return "no TLS"
}
```
In `main.go` we have:
```go
p.logger.Info(fmt.Sprintf("HTTP proxy server at %s%s with %s exiting with errors", p.config.Address, p.config.PrefixPath, mptls.Status(p.config.TLSConfig)), slog.String("error", err.Error()))
```

File: pkg/tls/validations.go [hunk 10]
```diff
@@ -0,0 +1,434 @@
+// Copyright (c) Abstract Machines
```
Comment (dborovcanin) [inline pkg/tls/validations.go:1]: Rename to `verifier`. Also, separate OCSP and CRL into subpackages.

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Replace `if-else` with `switch`.

File: pkg/tls/tls.go [hunk 11]
```diff
@@ -7,36 +7,141 @@ import (
...
 	}
-	return &tls.Config{
-		Certificates: []tls.Certificate{cert},
-		ClientAuth:   tls.RequireAndVerifyClientCert,
-		ClientCAs:    roots,
-	}, nil
+	return tlsConfig, nil
+}
+
+func (c *Config) Security() Security {
```
Comment (dborovcanin) [inline pkg/tls/tls.go:132]: Use this as a util function, do not use Config receiver. Also, move it to a different package.

File: pkg/tls/tls.go [hunk 12]
```diff
@@ -7,36 +7,141 @@ import (
...
+
+type Config struct {
+	CertFile     string `env:"CERT_FILE"                                  envDefault:""`
+	KeyFile      string `env:"KEY_FILE"                                   envDefault:""`
+	ServerCAFile string `env:"SERVER_CA_FILE"                             envDefault:""`
+	ClientCAFile string `env:"CLIENT_CA_FILE"                             envDefault:""`
+	Verifier     verifier.Config
+}
+
+func (c *Config) SecurityStatus() string {
```
Comment (dborovcanin) [inline pkg/tls/tls.go:59]: Same here, remove; use util method.

File: pkg/tls/tls.go [hunk 13]
```diff
@@ -7,36 +7,141 @@ import (
...
+	WithoutTLS Security = iota + 1
+	WithTLS
+	WithmTLS
+	WithmTLSVerify
 )
 
-// LoadTLSCfg return a TLS configuration that can be used in TLS servers.
-func LoadTLSCfg(ca, crt, key string) (*tls.Config, error) {
-	caCertPEM, err := os.ReadFile(ca)
+func (s Security) String() string {
```
Comment (dborovcanin) [inline pkg/tls/tls.go:36]: Remove, we do not need a new type; use the util method to determine log levels.

File: pkg/tls/tls.go [hunk 14]
```diff
@@ -7,36 +7,141 @@ import (
...
+		return "without TLS"
+	}
+}
+
+type Config struct {
+	CertFile     string `env:"CERT_FILE"                                  envDefault:""`
+	KeyFile      string `env:"KEY_FILE"                                   envDefault:""`
+	ServerCAFile string `env:"SERVER_CA_FILE"                             envDefault:""`
+	ClientCAFile string `env:"CLIENT_CA_FILE"                             envDefault:""`
+	Verifier     verifier.Config
```
Comment (dborovcanin) [inline pkg/tls/tls.go:56]: Replace with the list of verifiers, where Verifier is:
```go
type Verifier interface {
	// VerifyPeerCertificate...
	VerifyPeerCertificate(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error
}
```
and in the Config we have:

```go
func (c *Config) VerifyPeerCertificate(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
	for _, v := range c.Verifiers {
		if err := v.VerifyPeerCertificate(rawCerts, verifiedChains); err != nil {
			return err
		}
	}
	return nil
}
```
And you use different implementations of Verifier `ocsp` and `crl` just like you already did.

File: pkg/tls/tls.go [hunk 15]
```diff
@@ -7,36 +7,141 @@ import (
...
+	if len(clientCA) > 0 {
+		if tlsConfig.ClientCAs == nil {
+			tlsConfig.ClientCAs = x509.NewCertPool()
+		}
+		if !tlsConfig.ClientCAs.AppendCertsFromPEM(clientCA) {
+			return nil, errAppendCA
+		}
+		tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
+		if len(c.Verifier.ValidationMethods) > 0 {
+			tlsConfig.VerifyPeerCertificate = c.Verifier.VerifyPeerCertificate
```
Comment (dborovcanin) [inline pkg/tls/tls.go:126]: And here it would be something like:
```go
		tlsConfig.VerifyPeerCertificate = c.VerifyPeerCertificate
```

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Please fix CI liner comments.

**Review by dborovcanin — APPROVED** [review]

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: examples/server/mosquitto/server.conf [hunk 16]
```diff
@@ -0,0 +1,3 @@
+listener 1883
+listener 8000
+protocol websockets
```
Comment (dborovcanin) [inline examples/server/mosquitto/server.conf:3]: Please add `allow_anonymous true` to allow using more recent Mosquitto versions.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — APPROVED** [review]

**Pull request merged**